plugins/copr.py

```diff
@@ -212,25 +272,48 @@ def run(self):
             raise dnf.exceptions.Error(
                 _('Unknown subcommand {}.').format(subcommand))
 
-    def _list_installed_repositories(self, directory, enabled_only, disabled_only):
-        parser = ConfigParser()
+    def _list_repo_file(self, directory, file_name, enabled_only, disabled_only):
+        match_new = re.match("^copr:" + self.copr_hostname, file_name)
+        match_old = self.copr_url == self.default_url and re.match("^_copr_", file_name)
+        match_any = re.match("^copr:|^_copr_", file_name)
 
-        for root, dir, files in os.walk(directory):
-            for file_name in files:
-                if re.match("^_copr_", file_name):
-                    parser.read(directory + '/' + file_name)
+        if self.opts.hub:
+            if not match_new and not match_old:
+                return
+        elif not match_any:
+            return
+
+        parser = ConfigParser()
+        parser.read(directory + '/' + file_name)
 
         for copr in parser.sections():
             enabled = parser.getboolean(copr, "enabled")
             if (enabled and disabled_only) or (not enabled and enabled_only):
                 continue
 
-            copr_name = copr.split('-', 1)
-            msg = copr_name[0] + '/' + copr_name[1]
+            # repo ID has copr:<hub>:<user>:<project> format
+            if re.match("^copr:", copr):
+                copr_name = copr.rsplit(':', 2)
+                copr_hostname = copr_name[0].split(':', 1)[1]
+                msg = copr_hostname + '/' + copr_name[1] + '/' + copr_name[2]
+            # repo ID has <user>-<project> format, try to get hub from file name
+            elif re.match("^copr:", file_name):
+                copr_name = copr.split('-', 1)
+                copr_hostname = file_name.rsplit(':', 2)[0].split(':', 1)[1]
+                msg = copr_hostname + '/' + copr_name[0] + '/' + copr_name[1]
+            # no information about hub, assume the default one
+            else:
+                copr_name = copr.split('-', 1)
+                msg = self.default_hostname + '/' + copr_name[0] + '/' + copr_name[1]
             if not enabled:
                 msg += " (disabled)"
             print(msg)
 
+    def _list_installed_repositories(self, directory, enabled_only, disabled_only):
+        for root, dir, files in os.walk(directory):
+            for file_name in files:
```

Comment:
Instead of walking the directory and getting the filenames, you can simply use `self.base.repos` structure, which is a dictionary consisting of `copr_id: <repo object>` pairs and which is already provided by dnf itself. Just select repos that do have a key starting with `_copr:` (or `copr:` but having them prefixed with '_' is quite suitable here) and you can later recognize in `_list_repo_file` method whether they are enabled or not just by reading `<repo object>.enabled` attribute. This is much more simple and faster than parsing the files manually from the plugin and it also solves the above-mentioned problem with the `dnf copr list` command not correctly listing all the enabled copr repos when user just manually copy-pastes the repo configs from web UI.

Reply:
It obviously does help because we can just use the dictionary keys from `self.base.repos` as the source for hub ID.

Reply:
`.repofile` attribute can be read also (at the end of the method) but only for backward-compatibility reasons to support the old copr repo files that did not have copr ID in the required form (which would say: "hey, I am a copr repo") but they had a filename starting with `_copr_`. We can still list those copr repos but also display some kind of warning to stderr that will tell users that they can re-enable them to get rid of the warning. Later, we can drop this backward-compatibility logic (and warning) and keep only the one that relies on repo IDs. The reason for relying only on repo IDs is that we should be able to have all the (enabled) copr repos placed in just a single file and `dnf copr list` should still list them all. Placing all copr repos into a single file was already suggested here earlier but this is not something we need to be doing in the plugin. At the same time, `dnf copr list` should still work even if they are placed in the same file, otherwise what command displays is not a correct information about the system.

Reply:
Hmpfs, `dnf copr` should only operate with files added by `dnf copr`.  Doing otherwise would not only be kludge, but also it would be giving info that user didn't ask for.  E.g. if I write `dnf copr list`, I want to see what I added by `dnf copr enable` and nothing else.  Otherwise, neither repo ID nor repo file name has additional semantic.  Yes, there would be technically correct ways to achieve what you wish but I don't think it's really desired.

Reply:
> if I write dnf copr list, I want to see what I added by dnf copr enable and nothing else. 

+1

When I created this plugin originally. I decided to use `_copr` file prefix as no sane human will decide to use `_` as a prefix. I thought that this will immediately strike everybody that "do not touch this, this is kind of private file".  If you wget something or manually edit some file, then you are on your own. It may work or not. Not our business.